src/charm_state.py

```diff
@@ -16,6 +16,7 @@
     validator,
 )
 
+from database import ConnectionParams
```

Comment:
Perhaps this class should be defined here? Otherwise the state module depends on the database module and the database module depends on the state module?

Reply:
After the suggestion to use PebbleService, I changed the code a lot. Please take a look and feel free to reopen this comment or add new ones :)